libbpf-tools/profile.bpf.c

```diff
@@ -0,0 +1,99 @@
+/* SPDX-License-Identifier: GPL-2.0 */
+/* Copyright (c) 2022 LG Electronics */
+#include <vmlinux.h>
+#include <bpf/bpf_helpers.h>
+#include <bpf/bpf_core_read.h>
+#include <bpf/bpf_tracing.h>
+#include "profile.h"
+#include "maps.bpf.h"
+
+#define MAX_ENTRIES		10240
+
+const volatile bool kernel_stacks_only = false;
+const volatile bool user_stacks_only = false;
+const volatile bool include_idle = false;
+const volatile pid_t targ_pid = -1;
+const volatile pid_t targ_tid = -1;
+
+extern int LINUX_KERNEL_VERSION __kconfig;
+
+struct {
+	__uint(type, BPF_MAP_TYPE_STACK_TRACE);
+	__type(key, u32);
+} stackmap SEC(".maps");
+
+struct {
+	__uint(type, BPF_MAP_TYPE_HASH);
+	__type(key, struct key_t);
+	__type(value, sizeof(u64));
+	__uint(max_entries, MAX_ENTRIES);
+} counts SEC(".maps");
+
+SEC("perf_event")
+int do_perf_event(struct bpf_perf_event_data *ctx) {
+	u64 id = bpf_get_current_pid_tgid();
+	u32 pid = id >> 32;
+	u32 tid = id;
+	u64 *valp;
+	static const u64 zero;
+	struct key_t key = {};
+
+	if (!include_idle && tid == 0)
+		return 0;
+
+	if (targ_pid != -1 && targ_pid != pid)
+		return 0;
+	if (targ_tid != -1 && targ_tid != tid)
+		return 0;
+
+	key.pid = pid;
+	bpf_get_current_comm(&key.name, sizeof(key.name));
+
+	if (user_stacks_only)
+		key.kern_stack_id = -1;
+	else
+		key.kern_stack_id = bpf_get_stackid(&ctx->regs, &stackmap, 0);
+
+	if (kernel_stacks_only)
+		key.user_stack_id = -1;
+	else
+		key.user_stack_id = bpf_get_stackid(&ctx->regs, &stackmap, BPF_F_USER_STACK);
+
+	if (key.kern_stack_id >= 0) {
+		// populate extras to fix the kernel stack
+		extern bool CONFIG_ARM64 __kconfig __weak;
+
+		// arm64 only supported
```

Comment:
I wrote code to apply the same logic for x86 but to be honest: I do not understand what does this logic.
I saw it was present in `bcc` version of this gadget:
https://github.com/iovisor/bcc/blob/55dcd2ee5a6183c748c0dd553a9400bc7de768a1/tools/profile.py#L178

So, does someone know what do we need this code and what is the meaning of "fixing kernel stack"?

[Here is my code for information:

```patch
diff --git a/libbpf-tools/profile.bpf.c b/libbpf-tools/profile.bpf.c
index 08ac5687..9e4f5d0f 100644
--- a/libbpf-tools/profile.bpf.c
+++ b/libbpf-tools/profile.bpf.c
@@ -59,13 +59,13 @@ int do_perf_event(struct bpf_perf_event_data *ctx) {
 
        if (key.kern_stack_id >= 0) {
                // populate extras to fix the kernel stack
+               u64 ip = PT_REGS_IP(&ctx->regs);
+               u64 page_offset;
 
                // arm64 only supported
-#ifdef __TARGET_ARCH_arm64
+#if defined(__TARGET_ARCH_arm64)
                extern __u32 CONFIG_ARM64_VA_BITS __kconfig __weak;
                extern bool CONFIG_ARM64_64K_PAGES __kconfig __weak;
-               u64 ip = PT_REGS_IP(&ctx->regs);
-               u64 page_offset;
                u32 va;
 
                if (CONFIG_ARM64_VA_BITS)
@@ -79,11 +79,19 @@ int do_perf_event(struct bpf_perf_event_data *ctx) {
                        page_offset = (-(1UL << (va)));
                else
                        page_offset = (0xffffffffffffffffUL - (1UL << (va - 1)) + 1);
-
-               if (ip > page_offset) {
+#elif defined(__TARGET_ARCH_x86)
+#define __PAGE_OFFSET_BASE_L4 0xffff888000000000UL
+#define __PAGE_OFFSET_BASE_L5 0xff11000000000000UL
+               extern bool CONFIG_DYNAMIC_MEMORY_LAYOUT __kconfig __weak;
+               extern bool CONFIG_X86_5LEVEL __kconfig __weak;
+
+               if (CONFIG_DYNAMIC_MEMORY_LAYOUT && CONFIG_X86_5LEVEL)
+                       page_offset = __PAGE_OFFSET_BASE_L5;
+               else
+                       page_offset = __PAGE_OFFSET_BASE_L4;
+#endif /* __TARGET_ARCH_x86 */
+               if (ip > page_offset)
                        key.kernel_ip = ip;
-               }
-#endif
        }
 
        valp = bpf_map_lookup_or_try_init(&counts, &key, &zero);
```
]

Reply:
FYI, as I understand, It's to append ip address of the register to the top of the kernel stack trace to provide more detailed information to the user. page_offset is used to determine whether ip is a kernel address.

Thank you very much for the code for x86. But unfortunately I changed the code from yesterday...
because
- If the target doesn't have Kconfig, the profile won't work at all.
- On some versions of Linux, page_offset is not an effective value for this purpose.
Could you please check the updated code?
I'm really sorry and thank you.

Reply:
> I'm really sorry and thank you.

No problem, your actual version is simpler (and so better) than the previous and my proposal for x86.

Nonetheless, I do not really understand the goal of this code (or how it is printed).
Indeed, from my tests, I got the following output:

```
$ sudo ./profile -K 5
Sampling at 49 Hertz of all threads by kernel stack for 5 secs.
    #0  0xffffffff98ed69cb arch_local_irq_enable+0xb
    #1  0xffffffff99a00ec3 __schedule+0x773
    #2  0xffffffff99a0109f schedule+0x4f
    #3  0xffffffff98edf5f7 do_sched_yield+0x87
    #4  0xffffffff98edf62e __x64_sys_sched_yield+0xe
    #5  0xffffffff999f3b21 do_syscall_64+0x61
    #6  0xffffffff99c0007c entry_SYSCALL_64_after_hwframe+0x44
    kernel_ip#7  0xffffffff98ed69cb arch_local_irq_enable+0xb
...
```

You can see the entry where `k->kernel_ip` is not NULL begining with `kernel_ip`.
But this entry is the same as entry #0, so this is why I do not understand this code (maybe I am not using the tool like I should do).

Reply:
First of all, I fixed kernel_ip to be at the top of the call stack. Thank you so much for finding it.
And yes, as you said, I found a duplicate with the first entry in the call stack. I will check and update my opinion on it. I have ever seen this before and it's was produced with some conditions in my faint memory. please let me check it. Thank you.

Reply:
Where I see it before was a result of bcc docker. (https://github.com/iovisor/bcc/blob/master/QUICKSTART.md)
These are the test results for profile.py. And it has the same kernel_ip and call stack entry #0..
```
...
    do_syscall_64   // kernel_ip
    do_syscall_64   // call stack entry #0
    entry_SYSCALL_64_after_hwframe
    [unknown]
    [unknown]
    -                qemu-system-x86 (36165)
        7

    native_write_msr_safe  // kernel_ip
    native_write_msr_safe  // call stack entry #0
    kvm_set_shared_msr
    vmx_save_host_state
...
```
@yonghong-song 
If you have the same experience, could you please share your opinion on the equality of call stack entry #0 and kernel ip in profile.py?
Thank you.

Reply:
Hi.

This code was indeed already present in the bcc version of this tool, I just do not understand why we need this (maybe it was mandatory with some older kernels because we cannot get the full kernel stack from the `BPF_MAP_TYPE_STACK_TRACE`?).

Thank you for your investigation though :)

Best regards.